nvtabular/ops/categorify.py

```diff
@@ -954,16 +977,69 @@ def _write_uniques(dfs, base_path, col_selector: ColumnSelector, options: FitOpt
         # Construct gpu DataFrame from pyarrow data.
-        # `on_host=True` implies gpu-backed data.
+        # `on_host=True` implies gpu-backed data,
+        # because CPU-backed data would have never
+        # been converted from pandas to pyarrow.
         df = pa.concat_tables(dfs, promote=True)
-        df = dispatch._from_host(df)
+        if (
+            df.nbytes > options.cardinality_memory_limit
+            if options.cardinality_memory_limit
+            else False
+        ):
+            # Before fully converting this pyarrow Table
+            # to a cudf DatFrame, we can reduce the memory
+            # footprint of `df`. Since the size of `df`
+            # depends on the cardinality of the features,
+            # and NOT on the partition size, the remaining
+            # logic in this function has an OOM-error risk
+            # (even with tiny partitions).
+            size_columns = []
+            for col in col_selector.names:
+                name = col + "_size"
+                if name in df.schema.names:
+                    # Convert this column alone to cudf,
+                    # and drop the field from df. Note that
+                    # we are only converting this column to
+                    # cudf to take advantage of fast `max`
+                    # performance.
+                    size_columns.append(dispatch._from_host(df.select([name])))
+                    df = df.drop([name])
+                    # Use numpy to calculate the "minimum"
+                    # dtype needed to capture the "size" column,
+                    # and cast the type
+                    typ = np.min_scalar_type(size_columns[-1][name].max() * 2)
+                    size_columns[-1][name] = size_columns[-1][name].astype(typ)
+            # Convert the remaining columns in df to cudf,
+            # and append the type-casted "size" columns
+            df = dispatch._concat_columns([dispatch._from_host(df)] + size_columns)
+        else:
+            # Empty DataFrame - No need for type-casting
+            df = dispatch._from_host(df)
     else:
+        # For now, if we are not concatenating in host memory,
+        # we will assume that reducing the memory footprint of
+        # "size" columns is not a priority. However, the same
+        # type-casting optimization can also be done for both
+        # pandas and cudf-backed data here.
         df = _concat(dfs, ignore_index=True)
+
+    # Check if we should warn user that this Column is likely
+    # to cause memory-pressure issues
+    _df_size = df.memory_usage(deep=True, index=True).sum()
+    if (_df_size > options.cardinality_memory_limit) if options.cardinality_memory_limit else False:
+        warnings.warn(
+            f"Category DataFrame (with columns: {df.columns}) is {_df_size} "
+            f"bytes in size. This is large compared to the suggested "
+            f"upper limit of {options.cardinality_memory_limit} bytes!"
+        )
```

Comment:
I'm very open to feedback on this `UserWarning`. This PR introduces the `cardinality_memory_limit` argument as a mechanism to check if we should take extra steps to reduce memory usage.  However, I also assume it would be useful to warn the user if/when we are unable to keep the memory usage below this number (if the code eventually hits an OOM error, this warning should be a clear hint that a high-cardinality column is likely causing problems).

Reply:
It might make sense to add in the defaults are calculated (12.5% of host memory?)

 warnings.warn(
            f"Category DataFrame (with columns: {df.columns}) is {_df_size} "
            f"bytes in size. This is large compared to the suggested "
            f"upper limit of {options.cardinality_memory_limit} bytes!"
            f"(12.5% of host memory by default)"
        )